Integrate MediaProxy to bridge authenticated Matrix media (MSC3916)

tadzik opened the pull request:
Depends on https://github.com/matrix-org/matrix-appservice-bridge/pull/444

tadzik committed:
Integrate MediaProxy to bridge authenticated Matrix media (MSC3910)

turt2live commented:
just to confirm: does this mean to say MSC3916 instead of MSC3910? 3910 is a closed, but related, MSC.

tadzik commented:
> just to confirm: does this mean to say MSC3916 instead of MSC3910? 3910 is a closed, but related, MSC.

Yes, correct - I copied the old number from the MAB PR :)

heftig commented:
Is a release with this planned before Synapse defaults to disallowing unauthenticated media?

tadzik commented:
> Is a release with this planned before Synapse defaults to disallowing unauthenticated media?

Yes, that's the plan.

Half-Shot commented on src/config/BridgeConfig.ts:
```diff
@@ -13,7 +13,6 @@ export interface BridgeConfig {
     };
     homeserver: {
         url: string;
-        media_url?: string;
```
It would be kind somewhere to fail to start if this is still configured?

tadzik replied:
I'd keep it optional (maybe with a warning?) to make sure you can rollback easily without changing the configs too much.

We do require `mediaProxy` being configured now, so users will surely notice the change.

Half-Shot commented on src/bridge/IrcBridge.ts:
```diff
@@ -255,9 +260,10 @@ export class IrcBridge {
-            oldConfig.homeserver.media_url = newConfig.homeserver.media_url;
-            log.info(`Adjusted media_url to ${newConfig.homeserver.media_url}`);
+        if (JSON.stringify(oldConfig.ircService.mediaProxy) !== JSON.stringify(newConfig.ircService.mediaProxy)) {
+            await this._mediaProxy?.close();
```
Very nice!

Half-Shot commented on src/bridge/IrcBridge.ts:
```diff
@@ -209,6 +213,7 @@ export class IrcBridge {
             defaultTtlMs: 10 * 60 * 1000, // 10 mins
         });
         this.matrixBanSyncer = this.config.ircService.banLists && new MatrixBanSync(this.config.ircService.banLists);
+
```
rando space?

tadzik replied:
Dropped it in 395a8c1f0a848e57f09c5f81cd25a425840f7768

Half-Shot commented on config.sample.yaml:
```diff
@@ -607,6 +601,18 @@ ircService:
+  # Config for the media proxy, required to serve publically accessible URLs to authenticated Matrix media
+  mediaProxy:
+    # To generate a .jwk file:
+    # $ yarn -s ts-node src/generate-signing-key.ts > signingkey.jwk
```
Should we make a script entry for this to shortcut it a bit.

And is `ts-node` available by default?

tadzik replied:
Good point, `ts-node` is just a dev-dependency so far. Maybe we can even make it built-in, like with generating registration files (https://matrix-org.github.io/matrix-appservice-irc/latest/bridge_setup.html#4-registration)?

Half-Shot replied:
how chunky is it?

tadzik replied:
(We ended up converting it to raw JS).

Half-Shot reviewed (commented):
This looks good on the whole, only a few suggestions. Should we have tests though?

tadzik committed:
linting

tadzik committed:
Make singing key generator emit valid JSON; include a sample in test files

tadzik committed:
Fix media-related tests

tadzik committed:
Make mediaProxy required in the configuration

tadzik committed:
Await MediaProxy's startup

tadzik commented:
> This looks good on the whole, only a few suggestions. Should we have tests though?

We have some tests covering the case of "do these URLs look like MediaProxy's URLs" – I'm not sure how much further we want to take this, if it all.

Half-Shot commented:
> > This looks good on the whole, only a few suggestions. Should we have tests though?
> 
> We have some tests covering the case of "do these URLs look like MediaProxy's URLs" – I'm not sure how much further we want to take this, if it all.

I think that's sufficient, upstream has tests that the MP works?

Half-Shot commented on src/generate-signing-key.ts:
```diff
@@ -0,0 +1,11 @@
+import { webcrypto } from 'node:crypto';
```
To be honest, this could be a JavaScript file and save us a ts-node. We're not really even using any types here other than the Node API.

tadzik committed:
Add a test signingkey

tadzik committed:
De-typescriptify signing key generator

tadzik committed:
Update MAB dependency to pull in Media Proxy

tadzik committed:
Add mediaProxy config to e2e tests

tadzik committed:
Close the media proxy when killing the bridge

tadzik committed:
Make MediaProxy listen on random ports in tests

This reduces the chance of conflict with other processes,
and also stops one non-clean shutdown from blowing up the rest of the test suite.

tadzik committed:
Care less about mediaProxy not shutting down cleanly in kill()

tadzik committed:
Log the failure to close mediaproxy to not leave dangling rejections

tadzik committed:
Changelog

tadzik committed:
Merge branch 'develop' into tadzik/media-proxy

Half-Shot commented on spec/e2e/jest.config.js:
```diff
@@ -2,7 +2,7 @@
   preset: 'ts-jest',
   testEnvironment: 'node',
-  reporters: [['github-actions', {silent: false}], 'summary'],
+  // reporters: ['text', ['github-actions', {silent: false}], 'summary'],
```
oh no!

tadzik replied:
Oh no.

Half-Shot reviewed (approved):
I see no problems here. I have one concern about the test reporter :)

tadzik committed:
Undo accidental jest config changes

tadzik committed:
Remove the mention of HS url being used for media

tadzik committed:
Merge branch 'develop' into tadzik/media-proxy
